### DISCUSSION TITLE
ojph_expand changes image intensity

### DISCUSSION THREAD

**Discussion by stephenchenKL** [aous72/OpenJPH#133] [Q&A]
I use ojph_compress to compress a 16-bit grayscale tif image. The max intensity is 1600. After I expand the image back to tif image, its max intensity is changed to 51200. How to keep the image unchanged after compress-expand?

**Comment by aous72** [comment]
Hi Stephen,

Did you supply bit-depth? What was it?
I think TIFF support, added by @michaeldsmith, was to support the way the video industry use it.
It might be problematic for other uses -- let me know and I will look into making it more general way.

If your image is ppm, you should have no problem.

Cheers,
Aous.

**Comment by michaeldsmith** [comment]
@stephenchenKL can you please provide a sample file and the command-line you are using that causes this issue for you?  

The TIFF input code should just compress all 16bits exactly as provided, unless you provide a -bit_depth N command-line argument, which will cause openjph to only compress the top N bits in the input TIF file instead of using the file-bitdepth (usually 8 or 16).  Using 16bit TIFF input file with -bit_depth 10 or -bit_depth 12 is common for use in media industry applications where 10 or 12 bit image data is stored in the top 10 or 12 MSBs of a 16bit TIF file.

**Reply by stephenchenKL** [reply]
Hi @michaeldsmith and @aous72 , sorry for my late reply. I was busy with my project in the last two days. I figured out what's the problem. Actually, I did not use the TIFF to read images. Our software obtains a buffer holding data of a 16bit image, then compresses it with OpenJPH to a .jph file. I expand .jph file back to .tif file to display. 

A buffer_in class was added. The 8/10/12/16 bit auto-switching in this new class caused the problem. Because our buffer has a fixed 16 bit-depth. So I added a new class, handling 8-bit and 16-bit only:
      bytes_per_sample = bytes_p_smpl;
      max_val_num_bits = bytes_p_smpl*8;
      bit_depth[2] = bit_depth[1] = bit_depth[0] = max_val_num_bits;

Now it works fine with our image buffer.

Thank you so much for your replies. I really appreciate your explanation.

**Comment by aous72** [comment]
Hi Stephen,

Good to hear that you sorted it out.. Thank you for letting us know.

Kind regards,
Aous
